{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "40d5bd16-e455-4438-bb2e-c944bdba1b49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'In Python, custom exceptions are created by defining a new class that inherits from the built-in Exception class or one of its subclasses. The reason for using the Exception class as the base class for custom exceptions is to follow the established exception hierarchy in Python.\\n\\nBy inheriting from Exception (or one of its subclasses), you signal that your custom exception is part of the general exception hierarchy. This allows your custom exception to be caught along with other built-in exceptions in a unified manner. It also ensures that your custom exception can be used in generic except blocks without interfering with the handling of other exceptions.'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#ANSWER 1\n",
    "\n",
    "'''In Python, custom exceptions are created by defining a new class that inherits from the built-in Exception class or one of its subclasses. The reason for using the Exception class as the base class for custom exceptions is to follow the established exception hierarchy in Python.\n",
    "\n",
    "By inheriting from Exception (or one of its subclasses), you signal that your custom exception is part of the general exception hierarchy. This allows your custom exception to be caught along with other built-in exceptions in a unified manner. It also ensures that your custom exception can be used in generic except blocks without interfering with the handling of other exceptions.'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8b810cc0-f3df-4d1e-bf07-390c7e4469c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'BaseException'>\n",
      "  <class 'Exception'>\n",
      "    <class 'TypeError'>\n",
      "      <class 'decimal.FloatOperation'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'StopAsyncIteration'>\n",
      "    <class 'StopIteration'>\n",
      "    <class 'ImportError'>\n",
      "      <class 'ModuleNotFoundError'>\n",
      "      <class 'zipimport.ZipImportError'>\n",
      "    <class 'OSError'>\n",
      "      <class 'ConnectionError'>\n",
      "        <class 'BrokenPipeError'>\n",
      "        <class 'ConnectionAbortedError'>\n",
      "        <class 'ConnectionRefusedError'>\n",
      "        <class 'ConnectionResetError'>\n",
      "          <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'BlockingIOError'>\n",
      "      <class 'ChildProcessError'>\n",
      "      <class 'FileExistsError'>\n",
      "      <class 'FileNotFoundError'>\n",
      "      <class 'IsADirectoryError'>\n",
      "      <class 'NotADirectoryError'>\n",
      "      <class 'InterruptedError'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'PermissionError'>\n",
      "      <class 'ProcessLookupError'>\n",
      "      <class 'TimeoutError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'signal.itimer_error'>\n",
      "      <class 'socket.herror'>\n",
      "      <class 'socket.gaierror'>\n",
      "      <class 'ssl.SSLError'>\n",
      "        <class 'ssl.SSLCertVerificationError'>\n",
      "        <class 'ssl.SSLZeroReturnError'>\n",
      "        <class 'ssl.SSLWantWriteError'>\n",
      "        <class 'ssl.SSLWantReadError'>\n",
      "        <class 'ssl.SSLSyscallError'>\n",
      "        <class 'ssl.SSLEOFError'>\n",
      "      <class 'shutil.Error'>\n",
      "        <class 'shutil.SameFileError'>\n",
      "      <class 'shutil.SpecialFileError'>\n",
      "      <class 'shutil.ExecError'>\n",
      "      <class 'shutil.ReadError'>\n",
      "      <class 'urllib.error.URLError'>\n",
      "        <class 'urllib.error.HTTPError'>\n",
      "        <class 'urllib.error.ContentTooShortError'>\n",
      "      <class 'gzip.BadGzipFile'>\n",
      "    <class 'EOFError'>\n",
      "      <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "    <class 'RuntimeError'>\n",
      "      <class 'RecursionError'>\n",
      "      <class 'NotImplementedError'>\n",
      "        <class 'zmq.error.ZMQVersionError'>\n",
      "        <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "      <class '_frozen_importlib._DeadlockError'>\n",
      "      <class 'threading.BrokenBarrierError'>\n",
      "      <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "        <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "      <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "      <class 'pkg_resources.ExtractionError'>\n",
      "      <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "    <class 'NameError'>\n",
      "      <class 'UnboundLocalError'>\n",
      "    <class 'AttributeError'>\n",
      "      <class 'dataclasses.FrozenInstanceError'>\n",
      "    <class 'SyntaxError'>\n",
      "      <class 'IndentationError'>\n",
      "        <class 'TabError'>\n",
      "    <class 'LookupError'>\n",
      "      <class 'IndexError'>\n",
      "      <class 'KeyError'>\n",
      "        <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "        <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ValueError'>\n",
      "      <class 'UnicodeError'>\n",
      "        <class 'UnicodeEncodeError'>\n",
      "        <class 'UnicodeDecodeError'>\n",
      "        <class 'UnicodeTranslateError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'json.decoder.JSONDecodeError'>\n",
      "      <class 'ssl.SSLCertVerificationError'>\n",
      "      <class 'binascii.Error'>\n",
      "      <class '_hashlib.UnsupportedDigestmodError'>\n",
      "      <class 'calendar.IllegalMonthError'>\n",
      "      <class 'calendar.IllegalWeekdayError'>\n",
      "      <class 'dateutil.parser._parser.ParserError'>\n",
      "      <class 'pygments.util.ClassNotFound'>\n",
      "      <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "      <class 'email.errors.MessageDefect'>\n",
      "        <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "        <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "        <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "        <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "        <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "        <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "        <class 'email.errors.UndecodableBytesDefect'>\n",
      "        <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "        <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "        <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "        <class 'email.errors.HeaderDefect'>\n",
      "          <class 'email.errors.InvalidHeaderDefect'>\n",
      "          <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "          <class 'email.errors.NonPrintableDefect'>\n",
      "          <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "          <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "          <class 'email.errors.InvalidDateDefect'>\n",
      "      <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "      <class 'plistlib.InvalidFileException'>\n",
      "      <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "      <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "      <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "      <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "        <class 'pkg_resources.RequirementParseError'>\n",
      "      <class 'packaging.version.InvalidVersion'>\n",
      "    <class 'AssertionError'>\n",
      "    <class 'ArithmeticError'>\n",
      "      <class 'FloatingPointError'>\n",
      "      <class 'OverflowError'>\n",
      "      <class 'ZeroDivisionError'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.DivisionUndefined'>\n",
      "      <class 'decimal.DecimalException'>\n",
      "        <class 'decimal.Clamped'>\n",
      "        <class 'decimal.Rounded'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Inexact'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Subnormal'>\n",
      "          <class 'decimal.Underflow'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.FloatOperation'>\n",
      "        <class 'decimal.InvalidOperation'>\n",
      "          <class 'decimal.ConversionSyntax'>\n",
      "          <class 'decimal.DivisionImpossible'>\n",
      "          <class 'decimal.DivisionUndefined'>\n",
      "          <class 'decimal.InvalidContext'>\n",
      "    <class 'SystemError'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ReferenceError'>\n",
      "    <class 'MemoryError'>\n",
      "    <class 'BufferError'>\n",
      "    <class 'Warning'>\n",
      "      <class 'UserWarning'>\n",
      "        <class 'getpass.GetPassWarning'>\n",
      "        <class 'IPython.core.formatters.FormatterWarning'>\n",
      "      <class 'EncodingWarning'>\n",
      "      <class 'DeprecationWarning'>\n",
      "        <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "      <class 'PendingDeprecationWarning'>\n",
      "      <class 'SyntaxWarning'>\n",
      "      <class 'RuntimeWarning'>\n",
      "        <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "        <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "        <class 'pkg_resources.PEP440Warning'>\n",
      "      <class 'FutureWarning'>\n",
      "        <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "      <class 'ImportWarning'>\n",
      "      <class 'UnicodeWarning'>\n",
      "      <class 'BytesWarning'>\n",
      "      <class 'ResourceWarning'>\n",
      "      <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "      <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "    <class 'warnings._OptionError'>\n",
      "    <class 'runpy._Error'>\n",
      "    <class 're.error'>\n",
      "    <class 'sre_parse.Verbose'>\n",
      "    <class 'locale.Error'>\n",
      "    <class 'subprocess.SubprocessError'>\n",
      "      <class 'subprocess.CalledProcessError'>\n",
      "      <class 'subprocess.TimeoutExpired'>\n",
      "    <class 'tokenize.TokenError'>\n",
      "    <class 'tokenize.StopTokenizing'>\n",
      "    <class 'inspect.ClassFoundException'>\n",
      "    <class 'inspect.EndOfBlock'>\n",
      "    <class 'traitlets.traitlets.TraitError'>\n",
      "    <class 'copy.Error'>\n",
      "    <class 'concurrent.futures._base.Error'>\n",
      "      <class 'concurrent.futures._base.CancelledError'>\n",
      "      <class 'concurrent.futures._base.TimeoutError'>\n",
      "      <class 'concurrent.futures._base.InvalidStateError'>\n",
      "    <class 'socket._GiveupOnSendfile'>\n",
      "    <class 'struct.error'>\n",
      "    <class 'binascii.Incomplete'>\n",
      "    <class 'asyncio.exceptions.TimeoutError'>\n",
      "    <class 'asyncio.exceptions.InvalidStateError'>\n",
      "    <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "    <class 'asyncio.queues.QueueEmpty'>\n",
      "    <class 'asyncio.queues.QueueFull'>\n",
      "    <class '_queue.Empty'>\n",
      "    <class 'queue.Full'>\n",
      "    <class 'ctypes.ArgumentError'>\n",
      "    <class 'zmq.error.ZMQBaseError'>\n",
      "      <class 'zmq.error.ZMQError'>\n",
      "        <class 'zmq.error.ContextTerminated'>\n",
      "        <class 'zmq.error.Again'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'zmq.error.ZMQBindError'>\n",
      "      <class 'zmq.error.NotDone'>\n",
      "    <class '_pickle.PickleError'>\n",
      "      <class '_pickle.PicklingError'>\n",
      "      <class '_pickle.UnpicklingError'>\n",
      "    <class 'pickle._Stop'>\n",
      "    <class 'argparse.ArgumentError'>\n",
      "    <class 'argparse.ArgumentTypeError'>\n",
      "    <class 'traitlets.config.loader.ConfigError'>\n",
      "      <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "        <class 'traitlets.config.loader.ArgumentError'>\n",
      "      <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "    <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "      <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "    <class 'traitlets.config.application.ApplicationError'>\n",
      "    <class 'zlib.error'>\n",
      "    <class 'tornado.util.TimeoutError'>\n",
      "    <class '_curses.error'>\n",
      "    <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.KeyReuseError'>\n",
      "    <class 'tornado.gen.UnknownKeyError'>\n",
      "    <class 'tornado.gen.LeakedCallbackError'>\n",
      "    <class 'tornado.gen.BadYieldError'>\n",
      "    <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.Return'>\n",
      "    <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "    <class 'termios.error'>\n",
      "    <class '_lzma.LZMAError'>\n",
      "    <class 'shutil.RegistryError'>\n",
      "    <class 'shutil._GiveupOnFastCopy'>\n",
      "    <class 'configparser.Error'>\n",
      "      <class 'configparser.NoSectionError'>\n",
      "      <class 'configparser.DuplicateSectionError'>\n",
      "      <class 'configparser.DuplicateOptionError'>\n",
      "      <class 'configparser.NoOptionError'>\n",
      "      <class 'configparser.InterpolationError'>\n",
      "        <class 'configparser.InterpolationMissingOptionError'>\n",
      "        <class 'configparser.InterpolationSyntaxError'>\n",
      "        <class 'configparser.InterpolationDepthError'>\n",
      "      <class 'configparser.ParsingError'>\n",
      "        <class 'configparser.MissingSectionHeaderError'>\n",
      "    <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "    <class 'zipfile.BadZipFile'>\n",
      "    <class 'zipfile.LargeZipFile'>\n",
      "    <class 'entrypoints.BadEntryPoint'>\n",
      "    <class 'entrypoints.NoSuchEntryPoint'>\n",
      "    <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "    <class 'pydoc.ErrorDuringImport'>\n",
      "    <class 'executing.executing.NotOneValueFound'>\n",
      "    <class 'pure_eval.utils.CannotEval'>\n",
      "    <class 'pygments.util.OptionError'>\n",
      "    <class 'bdb.BdbQuit'>\n",
      "    <class 'pdb.Restart'>\n",
      "    <class 'pexpect.exceptions.ExceptionPexpect'>\n",
      "      <class 'pexpect.exceptions.EOF'>\n",
      "      <class 'pexpect.exceptions.TIMEOUT'>\n",
      "    <class 'ptyprocess.util.PtyProcessError'>\n",
      "    <class 'IPython.utils.process.FindCmdError'>\n",
      "    <class 'IPython.utils.path.HomeDirError'>\n",
      "    <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "    <class 'IPython.core.error.IPythonCoreError'>\n",
      "      <class 'IPython.core.error.TryNext'>\n",
      "      <class 'IPython.core.error.UsageError'>\n",
      "      <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'IPython.core.error.InputRejected'>\n",
      "    <class 'getopt.GetoptError'>\n",
      "    <class 'pygments.filters.ErrorToken'>\n",
      "    <class 'IPython.core.prefilter.PrefilterError'>\n",
      "    <class 'IPython.core.alias.AliasError'>\n",
      "      <class 'IPython.core.alias.InvalidAliasError'>\n",
      "    <class 'sqlite3.Error'>\n",
      "      <class 'sqlite3.InterfaceError'>\n",
      "      <class 'sqlite3.DatabaseError'>\n",
      "        <class 'sqlite3.InternalError'>\n",
      "        <class 'sqlite3.OperationalError'>\n",
      "        <class 'sqlite3.ProgrammingError'>\n",
      "        <class 'sqlite3.IntegrityError'>\n",
      "        <class 'sqlite3.DataError'>\n",
      "        <class 'sqlite3.NotSupportedError'>\n",
      "    <class 'sqlite3.Warning'>\n",
      "    <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "    <class 'xml.dom.DOMException'>\n",
      "      <class 'xml.dom.IndexSizeErr'>\n",
      "      <class 'xml.dom.DomstringSizeErr'>\n",
      "      <class 'xml.dom.HierarchyRequestErr'>\n",
      "      <class 'xml.dom.WrongDocumentErr'>\n",
      "      <class 'xml.dom.InvalidCharacterErr'>\n",
      "      <class 'xml.dom.NoDataAllowedErr'>\n",
      "      <class 'xml.dom.NoModificationAllowedErr'>\n",
      "      <class 'xml.dom.NotFoundErr'>\n",
      "      <class 'xml.dom.NotSupportedErr'>\n",
      "      <class 'xml.dom.InuseAttributeErr'>\n",
      "      <class 'xml.dom.InvalidStateErr'>\n",
      "      <class 'xml.dom.SyntaxErr'>\n",
      "      <class 'xml.dom.InvalidModificationErr'>\n",
      "      <class 'xml.dom.NamespaceErr'>\n",
      "      <class 'xml.dom.InvalidAccessErr'>\n",
      "      <class 'xml.dom.ValidationErr'>\n",
      "    <class 'prompt_toolkit.validation.ValidationError'>\n",
      "    <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "    <class 'prompt_toolkit.buffer._Retry'>\n",
      "    <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "    <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "    <class 'parso.parser.ParserSyntaxError'>\n",
      "    <class 'parso.parser.InternalParseError'>\n",
      "    <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "    <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "    <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "    <class 'jedi.inference.base_value.HasNoContext'>\n",
      "    <class 'jedi.inference.arguments.ParamIssue'>\n",
      "    <class 'jedi.api.exceptions._JediError'>\n",
      "      <class 'jedi.api.exceptions.InternalError'>\n",
      "      <class 'jedi.api.exceptions.WrongVersion'>\n",
      "      <class 'jedi.api.exceptions.RefactoringError'>\n",
      "    <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "    <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "    <class 'email.errors.MessageError'>\n",
      "      <class 'email.errors.MessageParseError'>\n",
      "        <class 'email.errors.HeaderParseError'>\n",
      "        <class 'email.errors.BoundaryError'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "      <class 'email.errors.CharsetError'>\n",
      "    <class 'uu.Error'>\n",
      "    <class 'http.client.HTTPException'>\n",
      "      <class 'http.client.NotConnected'>\n",
      "      <class 'http.client.InvalidURL'>\n",
      "      <class 'http.client.UnknownProtocol'>\n",
      "      <class 'http.client.UnknownTransferEncoding'>\n",
      "      <class 'http.client.UnimplementedFileMode'>\n",
      "      <class 'http.client.IncompleteRead'>\n",
      "      <class 'http.client.ImproperConnectionState'>\n",
      "        <class 'http.client.CannotSendRequest'>\n",
      "        <class 'http.client.CannotSendHeader'>\n",
      "        <class 'http.client.ResponseNotReady'>\n",
      "      <class 'http.client.BadStatusLine'>\n",
      "        <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'http.client.LineTooLong'>\n",
      "    <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "    <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "    <class 'psutil.Error'>\n",
      "      <class 'psutil.NoSuchProcess'>\n",
      "        <class 'psutil.ZombieProcess'>\n",
      "      <class 'psutil.AccessDenied'>\n",
      "      <class 'psutil.TimeoutExpired'>\n",
      "    <class 'psutil._pslinux._Ipv6UnsupportedError'>\n",
      "    <class 'tornado.queues.QueueEmpty'>\n",
      "    <class 'tornado.queues.QueueFull'>\n",
      "    <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "    <class 'xml.parsers.expat.ExpatError'>\n",
      "    <class 'xmlrpc.client.Error'>\n",
      "      <class 'xmlrpc.client.ProtocolError'>\n",
      "      <class 'xmlrpc.client.ResponseError'>\n",
      "      <class 'xmlrpc.client.Fault'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "    <class 'pkg_resources.ResolutionError'>\n",
      "      <class 'pkg_resources.VersionConflict'>\n",
      "        <class 'pkg_resources.ContextualVersionConflict'>\n",
      "      <class 'pkg_resources.DistributionNotFound'>\n",
      "      <class 'pkg_resources.UnknownExtra'>\n",
      "    <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "    <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "    <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n",
      "  <class 'GeneratorExit'>\n",
      "  <class 'SystemExit'>\n",
      "  <class 'KeyboardInterrupt'>\n",
      "  <class 'asyncio.exceptions.CancelledError'>\n",
      "  <class 'pkg_resources._vendor.more_itertools.more.AbortThread'>\n"
     ]
    }
   ],
   "source": [
    "#ANSWER 2\n",
    "\n",
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(\"  \" * indent + str(exception_class))\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 1)\n",
    "\n",
    "print_exception_hierarchy(BaseException)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa44edf8-2ad3-444c-ba73-3f4f13a6b9ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: division by zero\n"
     ]
    }
   ],
   "source": [
    "#ANSWER 3\n",
    "\n",
    "'''ArithmeticError is a base class for numeric errors that might occur during arithmetic operations. Two common errors defined in this class are:\n",
    "\n",
    "ZeroDivisionError: Raised when the second operand of a division or modulo operation is zero.'''\n",
    "\n",
    "\n",
    "try:\n",
    "    result = 5 / 0\n",
    "except ZeroDivisionError as e:\n",
    "    print(f\"Error: {e}\")\n",
    "#OverflowError: Raised when an arithmetic operation exceeds the limits of the data type.\n",
    "\n",
    "try:\n",
    "    result = 2 ** 1000  # OverflowError for exceeding integer limit\n",
    "except OverflowError as e:\n",
    "    print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "85223e9c-02b1-4a47-aecb-01a79ff1adad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: 'c'\n",
      "Error: list index out of range\n"
     ]
    }
   ],
   "source": [
    "#ANSWER 4\n",
    "\n",
    "#LookupError is a base class for errors that occur when a key or index is not found. Two common subclasses are:\n",
    "\n",
    "#KeyError: Raised when a dictionary key is not found.\n",
    "\n",
    "\n",
    "my_dict = {'a': 1, 'b': 2}\n",
    "try:\n",
    "    value = my_dict['c']\n",
    "except KeyError as e:\n",
    "    print(f\"Error: {e}\")\n",
    "#IndexError: Raised when a sequence subscript is out of range.\n",
    "\n",
    "\n",
    "my_list = [1, 2, 3]\n",
    "try:\n",
    "    value = my_list[4]\n",
    "except IndexError as e:\n",
    "    print(f\"Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b8c3f5f7-044a-4085-8a40-0c12a9bc6e3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ImportError: No module named 'non_existent_module'\n",
      "ModuleNotFoundError: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "#ANSWER 5\n",
    "\n",
    "#ImportError is a base class for exceptions raised when an import statement fails. It has a subclass called ModuleNotFoundError, which is more specific and raised when a module is not found during the import.\n",
    "\n",
    "\n",
    "try:\n",
    "    import non_existent_module\n",
    "except ImportError as e:\n",
    "    print(f\"ImportError: {e}\")\n",
    "\n",
    "try:\n",
    "    from non_existent_module import some_function\n",
    "except ModuleNotFoundError as e:\n",
    "    print(f\"ModuleNotFoundError: {e}\")\n",
    "#In the first example, an ImportError is caught when trying to import a non-existent module. In the second example, a more specific ModuleNotFoundError is caught when attempting to import a non-existent function from a module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0095149b-a3fd-49f8-81e6-9bee5d8fa923",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Specific Exception Handling: Catch specific exceptions rather than using a generic except block. This helps in better understanding and debugging.\\n\\nUse finally for Cleanup: If you need to ensure that certain code gets executed whether an exception occurs or not, use the finally block.\\n\\nAvoid Bare Excepts: Avoid using bare except without specifying the exception type. It can catch unexpected errors and make debugging difficult.\\n\\nHandle Exceptions Locally: Handle exceptions at the point where they occur, rather than letting them propagate too far up the call stack.\\n\\nLogging: Use the logging module to log exceptions and relevant information for debugging purposes.\\n\\nDon't Suppress Exceptions: Avoid suppressing exceptions without proper handling. If an exception occurs, it's usually a sign of a problem that needs attention.\\n\\nUse with Statement for Resources: When working with resources that need cleanup (like files or network connections), use the with statement to ensure proper resource management.\\n\\nCustom Exception Messages: Provide clear and informative error messages in your custom exceptions to aid in debugging.\\n\\nKeep Try Blocks Small: Limit the amount of code within a try block to minimize the chance of catching unintended exceptions.\\n\\nKnow the Built-in Exceptions: Be familiar with the built-in exceptions in Python to effectively handle common errors.\\n\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#ANSWER 6\n",
    "\n",
    "'''Specific Exception Handling: Catch specific exceptions rather than using a generic except block. This helps in better understanding and debugging.\n",
    "\n",
    "Use finally for Cleanup: If you need to ensure that certain code gets executed whether an exception occurs or not, use the finally block.\n",
    "\n",
    "Avoid Bare Excepts: Avoid using bare except without specifying the exception type. It can catch unexpected errors and make debugging difficult.\n",
    "\n",
    "Handle Exceptions Locally: Handle exceptions at the point where they occur, rather than letting them propagate too far up the call stack.\n",
    "\n",
    "Logging: Use the logging module to log exceptions and relevant information for debugging purposes.\n",
    "\n",
    "Don't Suppress Exceptions: Avoid suppressing exceptions without proper handling. If an exception occurs, it's usually a sign of a problem that needs attention.\n",
    "\n",
    "Use with Statement for Resources: When working with resources that need cleanup (like files or network connections), use the with statement to ensure proper resource management.\n",
    "\n",
    "Custom Exception Messages: Provide clear and informative error messages in your custom exceptions to aid in debugging.\n",
    "\n",
    "Keep Try Blocks Small: Limit the amount of code within a try block to minimize the chance of catching unintended exceptions.\n",
    "\n",
    "Know the Built-in Exceptions: Be familiar with the built-in exceptions in Python to effectively handle common errors.\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d74f067-9ed9-46e7-ad3e-794f248b2ed5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
